### ISSUE TITLE
Some text elements don't move with shapes

### ISSUE THREAD

**Issue by ryanmccool** [zsviczian/obsidian-excalidraw-plugin#538]
**Describe the bug**
Some text elements don't move with shapes.

**To Reproduce**
1. Go to 'any drawing that has shapes with embedded text'.
2. Grab all elements and move them to a different part of the drawing.
3. See error

Steps to fix:
1. Double click on text that didn't move.
2. Watch it auto-magically appear in the proper shape.

**Expected behavior**
My expectation is that text from the shapes move _with_ the shapes at time of dragging.

**Screenshots**
![image](https://user-images.githubusercontent.com/1365935/159064289-c4936d8f-d7dc-4d12-afef-852a46664890.png)

After grabbing all elements in a selector box and moving:
![image](https://user-images.githubusercontent.com/1365935/159064376-cde52974-f888-484c-9333-e47d1c3de468.png)

**Environment (please complete the following information):**
 - MacOS 12.3 and 12.2
 - Plugin version: 1.6.19 (and I've seen on a few versions before)
 - Obsidian version: v0.14.1 (at least) and v.0.14.2

**Additional context**
N/A.

**Comment by zsviczian** [comment]
I've seen this happen in the past, but I have always been unable to reproduce the error. Btw, I think the same error exists on excalidraw.com, i.e. this is likely not an error in the plugin, but in the core package.

- How do you select all? Using CMD+A, or by dragging the select box using the mouse?
- How did you create the elements?
  - Draw a circle, select the circle, press enter & enter the text - each time.
  - Or you created a circle with text and copied (CMD+C; CMD+V) that multiple times on the screen, then modified the text
  - Or you created a circle with text, and then used ALT+Drag to make copies?
  - Or ...?

Can you help me with two things:
1) experiment a bit to achieve the simplest situation when the error occurs and describe the exact steps to achieve the error.
2) try to see if you can reproduce the same error on Excalidraw.com

This is the test drawing I tried, but the error did not happen:
![image](https://user-images.githubusercontent.com/14358394/159114555-ce974118-9f80-450f-aaca-612be101c6f2.png)

**Comment by ryanmccool** [comment]
- How do you select all? Using CMD+A, or by dragging the select box using the mouse?
  **-  I typically always move items by using the selector box via the mouse. Subsequent tests have shown this the problem happens with either method of selection.**
- How did you create the elements?
  - Draw a circle, select the circle, press enter & enter the text - each time.
  - Or you created a circle with text and copied (CMD+C; CMD+V) that multiple times on the screen, then modified the text
  - Or you created a circle with text, and then used ALT+Drag to make copies?
     **- I've tried a few different ways and put those in my attached picture. After all of this further testing, I can't really say there is a hard pattern that I am finding.**

Here is after multiple move attempts and finally something happening:
<img width="775" alt="image" src="https://user-images.githubusercontent.com/1365935/159144530-978d0cb3-56a3-454b-9989-4c1ba49e511a.png">

I then went back to Obsidian and did a drag select of the bottom section and all items (they were all copy / pasted from above) showed the error:
<img width="738" alt="image" src="https://user-images.githubusercontent.com/1365935/159144640-5136d8d1-860d-4fda-be4f-ab2148230d52.png">

More tests:
I was able to reproduce in Obsidian again, but **never in the web version**. In the Obsidian version, the only thing that I could really tell was that an original set of shapes is what had the text mess up. By "original" I mean that after doing many copy / pastes and alt-drags, only the original set of shapes had the problem; the copies did not.

Sorry to say, this is proving quite difficult to pin down to a specific behavior!

**Comment by nicolasmarc1** [comment]
Hi! Saw the message i have a similar problem and i'm going to add more information when i can. But very quickly, i don't think this is a problem with core excalidraw but with how the text is attached to the shape/container and how it breaks when you copy under a unspecified way (maybe when a lot or successive fast copy? not sure when it happens yet) as it seems that if you look at the blockid of the new label in the json you can see that the json has in "originalText" and "text" the new version of the label but in "rawText" it has an older version of the text (the one from the element you copied).

I'm sending you an extract of the file with more info this weekend. Could you explain a bit more of how the json data and Text elements interact, especially when you make copy of elements? just to understand a bit better the plugin, thanks.

**Comment by zsviczian** [comment]
This is helpful, and gives me a starting point to investiage.
rawText is an Obsidian-only thing.

The difference between `text`, `originalText` and `rawText` is as follows:
- `rawText` is the text including markdown markup, exactly as the user has entered it
- `originalText` is the parsed version of `rawText`. If, for example, you have a markdown link with an alias in `rawText`, you will only see the alias in `originalText`. Well, to be exact, this is true for Excalidraw PREVIEW mode, when you switch to RAW mode `originalText` will also include the markdown markup.
- `text` is what you see on screen. In the case of a sticky note `text` will include linebreaks generated by word wrapping.

`# Text Elements` in the markdown file has a copy of `rawText`, each with a block reference, where the block reference is the element id in the JSON file.  The logic for this setup, is that Obsidian will keep the links up to date in the `# Text Elements` section, but not in the JSON file, thus `rawText` will be updated when for example you move files around and that changes the markdown links in the text. When you open Excalidraw the plugin updates text elements in the JSON based on the rawText taken from `# Text Elements`.

The additional complication is that in some cases the length of the parsed text is different from the length of the raw text. As such when you create sticky notes, I need to re-wrap the parsed text and update the element accordingly. This is where I will dig deeper to see if there is something in my code that is causing the disconnect between the container and the text.

**Comment by zsviczian** [comment]
I spent multiple hours today trying to replicate this issue. I wasn't able to.
Maybe this is because I am using a slightly newer build of the core excalidraw package... 
Please let me know if the issue happens again with 1.6.21.

**Comment by nicolasmarc1** [comment]
> I spent multiple hours today trying to replicate this issue. I wasn't able to. Maybe this is because I am using a slightly newer build of the core excalidraw package... Please let me know if the issue happens again with 1.6.21.

Sorry that you had to spend so much time on this issue. Yes it's difficult to reproduce on purpose as i don"t know exactly when/why it happens, I will post here if this happens again then. Also here's the extract of the file where you can see that the text don't move with shapes + when i tried to change the text to anonymize the file you can still see the original text in rawText.
[bug save Drawing 2022-03-17 00.15.57.excalidraw.md](https://github.com/zsviczian/obsidian-excalidraw-plugin/files/8381552/bug.save.Drawing.2022-03-17.00.15.57.excalidraw.md)

**Comment by zsviczian** [comment]
This is very helpful and I can see what is going on in the file. I just need to understand why...

The issue is with `JqdRr_qu6YPvXY0nuopEA`. When I process the during while saving it, I replace long ids with  short ones. So `qrxOLena` was originally `JqdRr_qu6YPvXY0nuopEA`. In this process I replace all occurrences of the long id with the new id.

I wonder why the replacement did not work in this case.

```json
{
	"type": "rectangle",
	"version": 562,
	"versionNonce": 440213923,
	"isDeleted": false,
	"id": "BnnR3p-9yj1wBg9a7-eOS",
	"fillStyle": "hachure",
	"strokeWidth": 1,
	"strokeStyle": "solid",
	"roughness": 1,
	"opacity": 100,
	"angle": 0,
	"x": 455.63892090188483,
	"y": -808.4750719827298,
	"strokeColor": "#000000",
	"backgroundColor": "transparent",
	"width": 252,
	"height": 325,
	"seed": 352611320,
	"groupIds": [],
	"strokeSharpness": "round",
	"boundElements": [
		{
			"id": "JqdRr_qu6YPvXY0nuopEA",
			"type": "text"
		}
	],
	"updated": 1648660850780,
	"link": null
},
{
	"type": "text",
	"version": 189,
	"versionNonce": 215580173,
	"isDeleted": false,
	"id": "qrxOLena",
	"fillStyle": "solid",
	"strokeWidth": 1,
	"strokeStyle": "dashed",
	"roughness": 1,
	"opacity": 100,
	"angle": 0,
	"x": 470.63892090188483,
	"y": -801.4750719827298,
	"strokeColor": "#000000",
	"backgroundColor": "#4c6ef5",
	"width": 242,
	"height": 35,
	"seed": 2093117320,
	"groupIds": [],
	"strokeSharpness": "round",
	"boundElements": [],
	"updated": 1648660850780,
	"link": null,
	"fontSize": 28,
	"fontFamily": 1,
	"text": "uuu",
	"rawText": "uuu",
	"baseline": 25,
	"textAlign": "center",
	"verticalAlign": "middle",
	"containerId": "BnnR3p-9yj1wBg9a7-eOS",
	"originalText": "uuu"
}
```

**Comment by nicolasmarc1** [comment]
ok so i did what you say manually (look for the id and container id of a broken element (find it by searching the blockid of the text element), in the container boundElement replace the long element id by the short one) in the original drawing (that have different ids because i copied and pasted in a new drawing instead of making a duplicate..., i thought it might have been why IDs were different but apparently i have the same problem in the original drawing) and now it works (the text moves with the shape)! 

So i think it confirms the problem. @ryanmccool can you also test this on your broken file to check if it's the same problem? 

Also @zsviczian i think there was maybe a word missing in "When I process the during while saving it, I replace long ids with short ones." so i'm not sure i understand why the IDs are replaced, is this to reduce the file size?

**Comment by zsviczian** [comment]
IDs are replaced to make them more user-friendly when you look at the markdown text under TextElements. Also, Excalidraw IDs by default include characters that are not allowed in Obsidian block references.

While I still wasn't able to reproduce the error... I made a small change in 1.6.22 which might have been causing this effect. Let's see. At least we are getting close to the root cause.

**Comment by zsviczian** [comment]
see also #569. I still can't find the root cause.
I have developed a fix for impacted files when loading the file.

**Comment by zsviczian** [comment]
I have deployed the fix with 1.6.23. I will close this issue now. Please raise a new one in case the issue surfaces again.
